storage: crash on debug keys

**Describe the problem**

While running `./cockroach debug keys` on a newly created roachprod cluster I found:
```
$ ./cockroach debug keys /mnt/data1/cockroach/
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1  a panic has occurred!
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +panic: iterator with constraint=2 is being used with key /Min that has constraint=1
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +(1) attached stack trace
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  -- stack trace:
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.gopanic
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/usr/local/go/src/runtime/panic.go:969
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).checkConstraint
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:318
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekGE
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:250
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.iterateOnReader
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:1059
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).MVCCIterate
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:713
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.runDebugKeys
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/debug.go:234
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:194
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).execute
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:852
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).ExecuteC
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).Execute
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.Run
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:315
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.doMain
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:121
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.Main
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:59
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | main.main
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach/main.go:26
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.main
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/usr/local/go/src/runtime/proc.go:204
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.goexit
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +Wraps: (2) panic: iterator with constraint=2 is being used with key /Min that has constraint=1
E210426 14:58:13.955666 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +Error types: (1) *withstack.withStack (2) *errutil.leafError
panic: iterator with constraint=2 is being used with key /Min that has constraint=1 [recovered]
	panic: iterator with constraint=2 is being used with key /Min that has constraint=1 [recovered]
	panic: iterator with constraint=2 is being used with key /Min that has constraint=1

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.RecoverAndReportPanic(0x5b6c7a0, 0xc0000760a8, 0xc000c1e000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:90 +0x9f
panic(0x44a6ea0, 0xc000d6fdd0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000dd8b80, 0x5b6c7a0, 0xc0000760a8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x2d3
panic(0x44a6ea0, 0xc000d6fdd0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).checkConstraint(0xc0004d5100, 0x0, 0x0, 0x0, 0x1c52e00)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:318 +0x2c5
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekGE(0xc0004d5100, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:250 +0x491
github.com/cockroachdb/cockroach/pkg/storage.iterateOnReader(0x5bda1a0, 0xc000d6fdb0, 0x0, 0x0, 0x0, 0x7e9f6ce, 0x2, 0x2, 0x0, 0xc0004c3b60, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine.go:1059 +0x249
github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).MVCCIterate(0xc0004d8750, 0x0, 0x0, 0x0, 0x7e9f6ce, 0x2, 0x2, 0x0, 0xc0004c3b60, 0x203000, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:713 +0x125
github.com/cockroachdb/cockroach/pkg/cli.runDebugKeys(0x7e75e60, 0xc0004c7d40, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/debug.go:234 +0x21f
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x7e75e60, 0xc0004c7d40, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:194 +0x79
github.com/spf13/cobra.(*Command).execute(0x7e75e60, 0xc0004c7d20, 0x1, 0x1, 0x7e75e60, 0xc0004c7d20)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:852 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x7e72c60, 0xc0000760a8, 0xc0004dc7aa, 0xa)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:315
github.com/cockroachdb/cockroach/pkg/cli.doMain(0x7e75e60, 0xc0004dc7aa, 0xa, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:121 +0x11a
github.com/cockroachdb/cockroach/pkg/cli.Main()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:59 +0x13a
main.main()
	/go/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach/main.go:26 +0x25
```

**To Reproduce**

```
roachprod create paul-test
roachprod stage paul-test cockroach
roachprod start paul-test
roachprod stop paul-test
```

Then on node 1: `./cockroach debug keys /mnt/data1/cockroach`.

**Expected behavior**
Not to crash.

**Environment:**
```
./cockroach version
Build Tag:        v21.1.0-alpha.3-2706-g0a0155ab6e
Build Time:       2021/04/26 12:39:53
Distribution:     CCL
Platform:         linux amd64 (x86_64-unknown-linux-gnu)
Go Version:       go1.15.11
C Compiler:       gcc 6.5.0
Build Commit ID:  0a0155ab6e8f4092a93c8dced865db33acff4cf4
Build Type:       release
```

--- a/pkg/cli/debug.go
+++ b/pkg/cli/debug.go
@@ -231,7 +231,7 @@ func runDebugKeys(cmd *cobra.Command, args []string) error {
 	}
 
 	results := 0
-	return db.MVCCIterate(debugCtx.startKey.Key, debugCtx.endKey.Key, storage.MVCCKeyAndIntentsIterKind, func(kv storage.MVCCKeyValue) error {
+	iterFunc := func(kv storage.MVCCKeyValue) error {
 		done, err := printer(kv)
 		if err != nil {
 			return err
@@ -244,7 +244,27 @@ func runDebugKeys(cmd *cobra.Command, args []string) error {
 			return iterutil.StopIteration()
 		}
 		return nil
-	})
+	}
+	endKey := debugCtx.endKey.Key
+	splitScan := false
+	// If the startKey is local and the endKey is global, split into two parts
+	// to do the scan. This is because MVCCKeyAndIntentsIterKind cannot span
+	// across the two key kinds.
+	if (len(debugCtx.startKey.Key) == 0 || keys.IsLocal(debugCtx.startKey.Key)) && !(keys.IsLocal(endKey) || bytes.Equal(endKey, keys.LocalMax)) {
+		splitScan = true
+		endKey = keys.LocalMax
+	}
+	if err := db.MVCCIterate(
+		debugCtx.startKey.Key, endKey, storage.MVCCKeyAndIntentsIterKind, iterFunc); err != nil {
+		return err
+	}
+	if splitScan {
+		if err := db.MVCCIterate(keys.LocalMax, debugCtx.endKey.Key, storage.MVCCKeyAndIntentsIterKind,
+			iterFunc); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 func runDebugBallast(cmd *cobra.Command, args []string) error {

